engine/src/main/java/com/datatorrent/stram/StreamingAppMasterService.java

```diff
@@ -1089,13 +1090,57 @@ private void finishApplication(FinalApplicationStatus finalStatus) throws YarnEx
    * Check for containers that were allocated in a previous attempt.
    * If the containers are still alive, wait for them to check in via heartbeat.
    */
-  private void previouslyAllocatedContainers(List<Container> containers)
+  private List<ContainerId> previouslyAllocatedContainers(List<Container> containersListByYarn)
   {
-    for (Container container : containers) {
-      this.allocatedContainers.put(container.getId().toString(), new AllocatedContainer(container));
-      //check the status
-      nmClient.getContainerStatusAsync(container.getId(), container.getNodeId());
+    List<ContainerId> containersToRelease = new ArrayList<>();
+
+    if (containersListByYarn.size() != 0) {
+
+      Map<String, Container> fromYarn = new HashMap<>();
+      Set<String> fromStreamingContainerManager = new HashSet<>();
+
+      LOG.debug("Containers list by YARN - {}", containersListByYarn);
+      LOG.debug("Containers list by Streaming Container Manger - {}", dnmgr.getPhysicalPlan().getContainers());
+
+      for (Container container : containersListByYarn) {
+        fromYarn.put(container.getId().toString(), container);
+      }
+
+      for (PTContainer ptContainer : dnmgr.getPhysicalPlan().getContainers()) {
+        // SCM starts the container without external ID.
+        if (ptContainer.getExternalId() != null) {
+          fromStreamingContainerManager.add(ptContainer.getExternalId());
+        }
+      }
+
+      for (PTContainer ptContainer : dnmgr.getPhysicalPlan().getContainers()) {
+
+        String containerId = ptContainer.getExternalId();
```

Comment:
It is possible that containerId is `null` here as well. It will be better to handle containers with ExternalId `null` in this loop only.